src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -32,33 +44,136 @@ pub struct SealedSstableBuilder {
+        tracker: Option<MemoryTracker>,
+    ) -> HummockResult<SealedSstableBuilder>;
+}
+
+/// Upload the in-memory SST data and metadata all at once.
+pub struct BatchUploadSealer {
+    sstable_store: SstableStoreRef,
+    policy: CachePolicy,
+}
+
+impl BatchUploadSealer {
+    pub fn new(sstable_store: SstableStoreRef, policy: CachePolicy) -> BatchUploadSealer {
+        Self {
+            sstable_store,
+            policy,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableBuilderSealer<InMemSstableWriter> for BatchUploadSealer {
+    async fn seal(
+        &self,
+        writer_output: <InMemSstableWriter as SstableWriter>::Output,
+        meta: SstableMeta,
+        sst_info: SstableInfo,
+        tracker: Option<MemoryTracker>,
+    ) -> HummockResult<SealedSstableBuilder> {
+        let bloom_filter_size = meta.bloom_filter.len();
+        let policy = self.policy;
+        let sstable_store = self.sstable_store.clone();
+        let upload_join_handle = tokio::spawn(async move {
+            let ret = sstable_store
+                .put_sst(sst_info.id, meta, writer_output, policy)
+                .await;
+            drop(tracker);
+            ret
+        });
+        Ok(SealedSstableBuilder {
+            sst_info,
+            upload_join_handle,
+            bloom_filter_size,
+        })
+    }
+}
+
+pub struct StreamingUploadSealer {
+    sstable_store: SstableStoreRef,
+}
+
+impl StreamingUploadSealer {
+    pub fn new(sstable_store: SstableStoreRef) -> Self {
+        Self { sstable_store }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableBuilderSealer<StreamingSstableWriter> for StreamingUploadSealer {
+    async fn seal(
+        &self,
+        writer_output: <StreamingSstableWriter as SstableWriter>::Output,
+        meta: SstableMeta,
+        sst_info: SstableInfo,
+        tracker: Option<MemoryTracker>,
+    ) -> HummockResult<SealedSstableBuilder> {
+        let sstable_store = self.sstable_store.clone();
+        let bloom_filter_size = meta.bloom_filter.len();
+        let upload_join_handle = tokio::spawn(async move {
+            let ret = sstable_store
+                .finish_put_sst_stream(writer_output, meta)
+                .await;
+            drop(tracker);
+            ret
+        });
+        Ok(SealedSstableBuilder {
+            sst_info,
+            upload_join_handle,
+            bloom_filter_size,
+        })
+    }
+}
+
 /// A wrapper for [`SstableBuilder`] which automatically split key-value pairs into multiple tables,
 /// based on their target capacity set in options.
 ///
 /// When building is finished, one may call `finish` to get the results of zero, one or more tables.
-pub struct CapacitySplitTableBuilder<F: TableBuilderFactory> {
-    /// When creating a new [`SstableBuilder`], caller use this closure to specify the id and
-    /// options.
+pub struct CapacitySplitTableBuilder<F, B, S>
+where
+    F: TableBuilderFactory<B>,
+    B: SstableWriterBuilder,
```

Comment:
This concept is confuse..

Reply:
We need to pass a `SstableWriter` into `SstableBuilder::new`, so `TableBuilderFactory` also needs a builder for `SstableWriter` to open a new `SstableWriter` for the new `SstableBuilder`.